Mathlib/Tactic/Linter/DeprecatedModule.lean

```diff
@@ -0,0 +1,146 @@
+/-
+Copyright (c) 2025 Damiano Testa. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Damiano Testa
+-/
+import Mathlib.Init
+
+/-!
+#  The `deprecated.module` linter
+
+The `deprecated.module` linter emits a warning when a file that has been renamed or split
+is imported.
+
+The usage is as follows. Write
+```lean
+import B
+...
+import Z
+
+deprecated_module since yyyy-mm-dd
+```
+in module `A` with the expectation that `A` contains nothing else.
+This triggers the `deprecated.module` linter to notify every file with `import A`
+to instead import the *direct imports* of `A`, that is `B, ..., Z`.
+-/
+
+open Lean Elab Command
+
+namespace Mathlib.Linter
+
+/--
+The `deprecated.module` linter emits a warning when a file that has been renamed or split
+is imported.
+The default value is `true`, since this linter is designed to warn projects downstream of `Mathlib`
+of refactors and deprecations in `Mathlib` itself.
+-/
+register_option linter.deprecated.module : Bool := {
+  defValue := true
+  descr := "enable the `deprecated.module` linter"
+}
+
+/--
+Defines the `deprecatedModuleExt` extension for adding a `HashSet` of pairs of
+* a module `Name` that has been deprecated and
+* an array of `Name`s of modules that should be imported instead
+
+to the environment.
+-/
+initialize deprecatedModuleExt :
+    SimplePersistentEnvExtension (Name × Array Name) (Std.HashSet (Name × Array Name)) ←
+  registerSimplePersistentEnvExtension {
+    addImportedFn := (·.foldl Std.HashSet.insertMany {})
+    addEntryFn := .insert
+  }
+
+/--
+`addModuleDeprecation` adds to the `deprecatedModuleExt` extension the pair consisting of the
+current module name and the array of its direct imports.
+
+It ignores the `Init` import, since this is a special module that is expected to be imported
+by all files.
+
+It also ignores the `Mathlib.Tactic.Linter.DeprecatedModule` import (namely, the current file),
+since there is no need to import this module.
+-/
+def addModuleDeprecation {m : Type → Type} [Monad m] [MonadEnv m] [MonadQuotation m] : m Unit := do
+  let modName ← getMainModule
+  modifyEnv (deprecatedModuleExt.addEntry ·
+    (modName, (← getEnv).imports.filterMap fun i ↦
+      if i.module == `Init ||
+         i.module == `Mathlib.Tactic.Linter.DeprecatedModule then none else i.module))
+
+/--
+`deprecated_module since yyyy-mm-dd` deprecates the current module `A` in favour of
```

Comment:
How about `deprecated_module (since := "2025-04-11")`, to match identically the `@[deprecated]` syntax. Fewer things to learn.

Reply:
I aligned the syntax to the `deprecated` one.